packages/service-library/src/servicelib/async_utils.py

```diff
@@ -0,0 +1,110 @@
+import asyncio
+from functools import wraps
+from collections import deque
+import attr
+from typing import List, Dict
+
+
+@attr.s
+class Context:
+    in_queue: asyncio.Queue = attr.ib()
+    out_queue: asyncio.Queue = attr.ib()
+    initialized: bool = attr.ib()
+
+
+def run_serialized_in_context(target_args: List[str] = None):
+    """All request to function with same calling context will be serialized.
+
+    Example:
+
+    Given the following decorated function
+
+        @run_serialized_in_context(target_args=["param3", "param1"])
+        async def func(param1, param2, param3):
+            await asyncio.sleep(1)
+
+    The context will be formed by the values of the arguments "param3" and "param1".
+    The values will must be serializable as they will be converted to string
+    and put together as storage key for the context.
+
+    The below calls will all run in a sequence:
+
+        functions = [
+            func(1, "something", 3),
+            func(1, "else", 3),
+            func(1, "here", 3),
+        ]
+        await asyncio.gather(*functions)
+
+    The following calls will run in parallel, because they have different contexts:
+
+        functions = [
+            func(1, "something", 3),
+            func(2, "else", 3),
+            func(3, "here", 3),
+        ]
+        await asyncio.gather(*functions)
+
+    """
+    target_args = [] if target_args is None else target_args
+
+    def internal(decorated_function):
+        contexts = {}
+
+        def get_context(args, kwargs: Dict) -> Context:
+            search_args = dict(zip(decorated_function.__code__.co_varnames, args))
```

Comment:
careful with these __variables__ remember what happened with coro.__name__ ?
can't you use inspect maybe [here](https://docs.python.org/3.3/library/inspect.html)

Reply:
I've abandoned an inspect based solution in favour of this one because I was not able to retrive the entire list of arguments (their names).

I was trying [to use this](https://stackoverflow.com/a/218709/2855718). Ended up [doing this](https://stackoverflow.com/a/218625/2855718) partially. Will push new changes which address this issue.

The implementation depends on CPython's internals. Since I don't see us moving to PyPy or other alternative Python interpreters, I will push a proper solution.